README.md
# PBKDF2ServiceProvider #

This service provider for __[Silex](http://silex-project.org)__ enables developers to seamlessly leverage the
[Password-Based Key Derivation Function (PBKDF2)](http://www.ietf.org/rfc/rfc2898.txt) in their web applications.
It can automatically choose between a pure-PHP implementation of the algorithm or, if available in the `hash`
extension, the C-based function `hash_pbkdf2()` proposed in [PHP #60813](https://bugs.php.net/bug.php?id=60813).


## Getting started ##

Using this service provider is easy, all you need to do is register its namespace in the autoloader stack and
register an instance of the service provider in the silex application:

``` php
<?php
/* ... */
$app['autoloader']->registerNamespaces(array(
    'Nrk\PBKDF2' => __DIR__.'/../vendor/PBKDF2ServiceProvider/lib',
));

$app->register(new Nrk\PBKDF2\PBKDF2ServiceProvider(), array(
    'pbkdf2.salt' => 'my_salt_string',
));

$app->get('/', function(Silex\Application $app) {
    $key = $app['pbkdf2']('my_password');
});
```

If you are using [Composer](http://getcomposer.org/) to manage the dependencies of your Silex application (which
is highly recommended anyway) you do not need to register the namespace in the autoloader stack since this step
is managed for you by Composer itself.

The only required parameter when registering the service provider instance is `pbkdf2.salt` which stores the
salt string used on each supplied password to calculate the resulting key. Optionally `pbkdf2.salt` can be
omitted, but then you must provide a salt string as the second parameter of the generator method. This can
be useful when you want to use dynamically generated salt strings for each password.

``` php
$key = $app['pbkdf2']('my_password', 'my_salt_string');
```

This is the full list of customizable parameters supported by PBKDF2ServiceProvider:

- `pbkdf2.salt`: common salt string used for each password.
- `pbkdf2.algorithm`: hashing algorithm used to generate the key [default: `sha1`].
- `pbkdf2.key_length`: length in bytes of the resulting key [default: `160`].
- `pbkdf2.iterations`: number of hash iterations performed on the password and salt [default: `1000`].
- `pbkdf2.function`: custom function used to generate the key [default: pure-PHP function or `hash_pbkdf2()` if available].


## Testing ##

In order to be able to run the test suite of the provider you must download [Composer](http://packagist.org/about-composer)
in the root of the repository and then install the needed dependencies.

```bash
$ wget http://getcomposer.org/composer.phar
$ php composer.phar install
$ phpunit
```


## Dependencies ##

- PHP >= 5.3.2


## Project links ##
- [Source code](http://github.com/nrk/PBKDF2ServiceProvider)
- [Issue tracker](http://github.com/nrk/PBKDF2ServiceProvider/issues)


## Author ##

- [Daniele Alessandri](mailto:suppakilla@gmail.com) ([twitter](http://twitter.com/JoL1hAHN))


## License ##

The code for PBKDF2ServiceProvider is distributed under the terms of the __MIT license__ (see LICENSE).


CHANGELOG
v0.1.0 (2012-xx-xx)
  * Initial release.


tests/bootstrap.php
<?php

/*
 * This file is part of the PredisServiceProvider package.
 *
 * (c) Daniele Alessandri <suppakilla@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

require_once __DIR__.'/../vendor/.composer/autoload.php';
require_once __DIR__.'/../lib/Nrk/PBKDF2/PBKDF2ServiceProvider.php';


tests/Nrk/PBKDF2/PBKDF2ServiceProvider.php
<?php

/*
 * This file is part of the PredisServiceProvider package.
 *
 * (c) Daniele Alessandri <suppakilla@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Nrk\PBKDF2\Silex;

use Silex\Application;
use Silex\ServiceProviderInterface;
use Nrk\PBKDF2\PBKDF2ServiceProvider as PBKDF2;

class PBKDF2ServiceProviderTest extends \PHPUnit_Framework_TestCase
{
    /**
     * Returns an initialized Silex application with a registered instance
     * of PBKDF2ServiceProvider.
     *
     * @param Array $providerConfiguration Parameters for the provider configuration.
     * @return Application
     */
    protected function getApplication(Array $providerConfiguration = array())
    {
        $app = new Application();
        $app->register(new PBKDF2(), $providerConfiguration);

        return $app;
    }

    /**
     *
     */
    public function testDefaultConfiguration()
    {
        $expected = "6ab15f4d10ae4d40667ed22574f3d645d69382d842bfa55f1c55343ef61df635ccc173b723524ffa64f3a26688450e5a77cb".
                    "de23b4c63ca80e502f1cddc3bdff03d256bccd1d42de715204dfddfd1638d471c0534f13b906ec9529c3a789b6185d728e6b".
                    "c47a56bc872abc25ea5be8fc3f4e564d53c6a3633c1891f0a0bf6e60c5912a87e6e69bcfe90e8cc83094dc1640a69642e53b".
                    "b410ed305e2be62a93b7";

        $app = $this->getApplication(array(
            'pbkdf2.salt' => $salt = 'salt_string',
        ));

        $this->assertSame($salt, $app['pbkdf2.salt']);

        $this->assertSame(PBKDF2::ALGORITHM, $app['pbkdf2.algorithm']);
        $this->assertSame(PBKDF2::KEY_LENGTH, $app['pbkdf2.key_length']);
        $this->assertSame(PBKDF2::ITERATIONS, $app['pbkdf2.iterations']);

        $this->assertSame($expected, $app['pbkdf2']('password'));
    }

    /**
     *
     */
    public function testSaltStringAsArgument()
    {
        $expected = "6ab15f4d10ae4d40667ed22574f3d645d69382d842bfa55f1c55343ef61df635ccc173b723524ffa64f3a26688450e5a77cb".
                    "de23b4c63ca80e502f1cddc3bdff03d256bccd1d42de715204dfddfd1638d471c0534f13b906ec9529c3a789b6185d728e6b".
                    "c47a56bc872abc25ea5be8fc3f4e564d53c6a3633c1891f0a0bf6e60c5912a87e6e69bcfe90e8cc83094dc1640a69642e53b".
                    "b410ed305e2be62a93b7";

        $app = $this->getApplication();

        $this->assertSame($expected, $app['pbkdf2']('password', 'salt_string'));
    }

    /**
     * @expectedException RuntimeException
     * @expectedExceptionMessage Please configure a salt string in "pbkdf2.salt" or pass its value as the second argument.
     */
    public function testUnspecifiedSaltStringThrowsException()
    {
        $app = $this->getApplication();
        $app['pbkdf2']('password');
    }

    /**
     *
     */
    public function testAlgorithmConfiguration()
    {
        $app = $this->getApplication(array(
            'pbkdf2.salt' => 'salt_string',
            'pbkdf2.algorithm' => $algorithm = 'sha256',
        ));

        $this->assertSame($algorithm, $app['pbkdf2.algorithm']);
        $this->assertSame(PBKDF2::KEY_LENGTH * 2, strlen($app['pbkdf2']('password')));
    }

    /**
     * @expectedException InvalidArgumentException
     * @expectedExceptionMessage Unknown hashing algorithm: foo.
     */
    public function testUnsupportedAlgorithmThrowsException()
    {
        $app = $this->getApplication(array(
            'pbkdf2.salt' => 'salt_string',
            'pbkdf2.algorithm' => 'foo',
        ));
    }

    /**
     *
     */
    public function testKeyLengthConfiguration()
    {
        $app = $this->getApplication(array(
            'pbkdf2.salt' => 'salt_string',
            'pbkdf2.key_length' => $length = 256,
        ));

        $this->assertSame($length, $app['pbkdf2.key_length']);
        $this->assertSame($length * 2, strlen($app['pbkdf2']('password')));
    }

    /**
     *
     */
    public function testIterationsConfiguration()
    {
        $app = $this->getApplication(array(
            'pbkdf2.salt' => 'salt_string',
            'pbkdf2.iterations' => $iterations = 2000,
        ));

        $this->assertSame($iterations, $app['pbkdf2.iterations']);
        $this->assertSame(PBKDF2::KEY_LENGTH * 2, strlen($app['pbkdf2']('password')));
    }

    /**
     *
     */
    public function testCustomFunctionConfiguration()
    {
        $expected = "6ab15f4d10ae4d40667ed22574f3d645d69382d842bfa55f1c55343ef61df635ccc173b723524ffa64f3a26688450e5a77cb".
                    "de23b4c63ca80e502f1cddc3bdff03d256bccd1d42de715204dfddfd1638d471c0534f13b906ec9529c3a789b6185d728e6b".
                    "c47a56bc872abc25ea5be8fc3f4e564d53c6a3633c1891f0a0bf6e60c5912a87e6e69bcfe90e8cc83094dc1640a69642e53b".
                    "b410ed305e2be62a93b7";

        $callable = $this->getMock('stdClass', array('__invoke'));
        $callable->expects($this->once())
                 ->method('__invoke')
                 ->with(PBKDF2::ALGORITHM, 'password', 'salt_string', PBKDF2::ITERATIONS, PBKDF2::KEY_LENGTH)
                 ->will($this->returnValue($expected));

        $app = $this->getApplication(array(
            'pbkdf2.salt' => 'salt_string',
            'pbkdf2.function' => $callable,
        ));

        $this->assertSame($callable, $app['pbkdf2.function']);
        $this->assertSame($expected, $app['pbkdf2']('password'));
    }

    /**
     * @expectedException InvalidArgumentException
     * @expectedExceptionMessage The value of "pbkdf2.function" must be a valid callable object.
     */
    public function testInvalidCustomFunctionThrowsException()
    {
        $app = $this->getApplication(array(
            'pbkdf2.salt' => 'salt_string',
            'pbkdf2.function' => new \stdClass(),
        ));
    }
}


lib/Nrk/PBKDF2/PBKDF2ServiceProvider.php
<?php

/*
 * This file is part of the PBKDF2ServiceProvider package.
 *
 * (c) Daniele Alessandri <suppakilla@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Nrk\PBKDF2;

use Silex\Application;
use Silex\ServiceProviderInterface;

/**
 * Exposes the Password-Based Key Derivation Function (PBKDF2) to Silex applications
 * using a pure-PHP implementation of the algorithm.
 *
 * This service provider can automatically switch to hash_pbkdf2() if this function
 * has been compiled for the hash extension as proposed in PHP request #60813.
 *
 * @author Daniele Alessandri <suppakilla@gmail.com>
 * @link http://www.ietf.org/rfc/rfc2898.txt
 * @link https://bugs.php.net/bug.php?id=60813
 */
class PBKDF2ServiceProvider implements ServiceProviderInterface
{
    const ALGORITHM = 'sha1';
    const KEY_LENGTH = 160;
    const ITERATIONS = 1000;

    /**
     * Returns the default hashing function detecting if hash_pbkdf2() has been compiled
     * and is available in the hash extension. The hashing function returned by this method
     * must be a callable object compatible with the following signature:
     *
     * function($algorithm, $password, $salt, $iterations, $length)
     *
     * @return mixed
    */
    protected function getHashFunction()
    {
        if (function_exists('hash_pbkdf2')) {
            return function($algorithm, $password, $salt, $iterations, $length) {
                return hash_pbkdf2($algorithm, $password, $salt, $iterations, $length);
            };
        }

        return function($algorithm, $password, $salt, $iterations, $length) {
            $derivedKey = '';

            for ($blockPos = 1; $blockPos < $length; $blockPos++) {
                $block = $hmac = hash_hmac($algorithm, $salt . pack('N', $blockPos), $password, true);
                for ($i = 1; $i < $iterations; $i++) {
                    $block ^= ($hmac = hash_hmac($algorithm, $hmac, $password, true));
                }
                $derivedKey .= $block;
            }

            return bin2hex(substr($derivedKey, 0, $length));
        };
    }

    /**
     * {@inheritdoc}
     */
    public function register(Application $app)
    {
        if (isset($app['pbkdf2.function'])) {
            if (!is_callable($app['pbkdf2.function'])) {
                throw new \InvalidArgumentException('The value of "pbkdf2.function" must be a valid callable object.');
            }
        }
        else {
            $app['pbkdf2.function'] = $app->protect($this->getHashFunction($app));
        }

        if (isset($app['pbkdf2.algorithm'])) {
            if (!in_array($algorithm = $app['pbkdf2.algorithm'], hash_algos())) {
                throw new \InvalidArgumentException("Unknown hashing algorithm: $algorithm.");
            }
        }
        else {
            $app['pbkdf2.algorithm'] = self::ALGORITHM;
        }

        if (!isset($app['pbkdf2.key_length'])) {
            $app['pbkdf2.key_length'] = self::KEY_LENGTH;
        }

        if (!isset($app['pbkdf2.iterations'])) {
            $app['pbkdf2.iterations'] = self::ITERATIONS;
        }

        $app['pbkdf2'] = $app->protect(function($password, $salt = null) use($app) {
            if ($salt === null) {
                if (!isset($app['pbkdf2.salt'])) {
                    throw new \RuntimeException('Please configure a salt string in "pbkdf2.salt" or pass its value as the second argument.');
                }
                $salt = $app['pbkdf2.salt'];
            }

            $arguments = array(
                $app['pbkdf2.algorithm'],
                $password,
                $salt,
                (int) $app['pbkdf2.iterations'],
                (int) $app['pbkdf2.key_length'],
            );

            return call_user_func_array($app['pbkdf2.function'], $arguments);
        });
    }
}
